### PR TITLE
Snapshot metadata

### PR DESCRIPTION
**Pull request by DoctorVoid**

### CODE REVIEW

File: src/plugins/snapshot/snapshot-listener.ts [hunk 1]
```diff
@@ -44,61 +55,65 @@ export class SnapshotListener implements GraphQLRequestListener<PolarisGraphQLCo
...
-                delete parsedResult.extensions.prefetchBuffer;
-                const snapshotPage = new SnapshotPage(JSON.stringify(parsedResult));
-                await snapshotRepository.save({} as any, snapshotPage);
-                pagesIds.push(snapshotPage.getId());
-                context.snapshotContext!.startIndex! += context.snapshotContext!.countPerPage!;
-                currentPageIndex++;
-            } while (
-                currentPageIndex <
-                context.snapshotContext!.totalCount! / context.snapshotContext!.countPerPage!
+            context = cloneDeep(context);
```
Comment (chenshoo) [inline src/plugins/snapshot/snapshot-listener.ts]: why?

File: src/plugins/snapshot/snapshot-listener.ts [hunk 2]
```diff
@@ -44,61 +55,65 @@ export class SnapshotListener implements GraphQLRequestListener<PolarisGraphQLCo
...
+                    snapshotPages.push(new SnapshotPage(uuid));
+                    return uuid;
+                });
+            await snapshotRepository.save({} as any, snapshotPages);
+            const irrelevantEntitiesOfPages: IrrelevantEntitiesResponse[] = [];
+            snapshotMetadata.setPageIds(pagesIds);
+            snapshotMetadata.setDataVersion(context.returnedExtensions.globalDataVersion);
+            snapshotMetadata.setTotalCount(context.snapshotContext?.totalCount!);
+            snapshotMetadata.setPagesCount(pageCount);
+            await snapshotMetadataRepository.save({} as any, snapshotMetadata);
```
Comment (chenshoo) [inline src/plugins/snapshot/snapshot-listener.ts:96]: extract init of snapshot metadata to a new function

File: src/plugins/snapshot/snapshot-listener.ts [hunk 3]
```diff
@@ -121,4 +136,129 @@ export class SnapshotListener implements GraphQLRequestListener<PolarisGraphQLCo
...
+            snapshotRepository,
+            snapshotMetadataRepository,
+            snapshotMetadata,
+            snapshotPages[currentPageIndex],
+            irrelevantEntitiesOfPages,
+        );
+        context.snapshotContext!.startIndex! += context.snapshotContext!.countPerPage!;
+        ++currentPageIndex;
+        while (currentPageIndex < pageCount) {
+            const clonedContext = cloneDeep(context);
```
Comment (chenshoo) [inline src/plugins/snapshot/snapshot-listener.ts]: why here too

File: src/server/polaris-server.ts [hunk 4]
```diff
@@ -61,7 +63,34 @@ export class PolarisServer {
                     config.connectionManager as PolarisConnectionManager,
                 ).getRepository(SnapshotPage);
                 const result = await snapshotRepository.findOne({} as any, id);
-                res.send(result?.getData());
+                if (!result) {
+                    res.send({});
+                } else {
+                    result.setLastAccessedTime(new Date());
```
Comment (chenshoo) [inline src/server/polaris-server.ts]: you don't need this, you have annotation on this

File: src/server/polaris-server.ts [hunk 5]
```diff
@@ -61,7 +63,34 @@ export class PolarisServer {
...
+                    await snapshotRepository.save({} as any, result);
+                    const responseToSend =
+                        result!.getStatus() !== SnapshotStatus.DONE
+                            ? { status: result!.getStatus(), id: result!.getId() }
+                            : result!.getData();
+                    res.send(responseToSend);
+                }
+            });
+
+            app.get('/snapshot/metadata', async (req: express.Request, res: express.Response) => {
```
Comment (chenshoo) [inline src/server/polaris-server.ts]: seperate this from polaris server

File: src/server/polaris-server.ts [hunk 6]
```diff
@@ -61,7 +63,34 @@ export class PolarisServer {
...
+                const realityHeader: string | string[] | undefined = req.headers[REALITY_ID];
+                const realityId: number = realityHeader ? +realityHeader : 0;
+                const snapshotMetadataRepository = getConnectionForReality(
+                    realityId,
+                    this.polarisServerConfig.supportedRealities as any,
+                    config.connectionManager as PolarisConnectionManager,
+                ).getRepository(SnapshotMetadata);
+                const result = await snapshotMetadataRepository.findOne({} as any, id);
+                if (result) {
+                    result.setLastAccessedTime(new Date());
```
Comment (chenshoo) [inline src/server/polaris-server.ts]: here too

File: test/integration/server/schema/resolvers.ts [hunk 7]
```diff
@@ -33,7 +37,8 @@ export const resolvers = {
             const connection = getPolarisConnectionManager().get();
             polarisGraphQLLogger.debug("I'm the resolver of all books", context);
             return {
-                getData: (startIndex?: number, pageSize?: number): Promise<Book[]> => {
+                getData: async (startIndex?: number, pageSize?: number): Promise<Book[]> => {
+                    // await sleep(15000);
```
Comment (chenshoo) [inline test/integration/server/schema/resolvers.ts]: delete sleep

File: test/integration/server/utils/logger.ts [hunk 8]
```diff
@@ -17,7 +17,7 @@ const applicationLogProperties = {
 
 export const polarisGraphQLLogger = new PolarisGraphQLLogger(
     {
-        loggerLevel: 'info',
+        loggerLevel: 'debug',
```
Comment (chenshoo) [inline test/integration/server/utils/logger.ts]: change back to info

File: src/plugins/extensions/extensions-listener.ts [hunk 9]
```diff
@@ -22,7 +22,11 @@ export class ExtensionsListener implements GraphQLRequestListener<PolarisGraphQL
             if (!this.shouldAddWarningsToExtensions) {
                 context.returnedExtensions.warnings = undefined;
             }
-            response.extensions = { ...response.extensions, ...context.returnedExtensions };
+            const extensionsToReturn = { ...response.extensions, ...context.returnedExtensions };
+            if (context.snapshotContext && extensionsToReturn.prefetchBuffer) {
+                delete extensionsToReturn.prefetchBuffer;
```
Comment (osher-sade) [inline src/plugins/extensions/extensions-listener.ts:27]: Don't use delete on an existing object, return a new object that doesn't include this field

File: src/plugins/snapshot/snapshot-listener.ts [hunk 10]
```diff
@@ -44,61 +55,65 @@ export class SnapshotListener implements GraphQLRequestListener<PolarisGraphQLCo
...
+            context = cloneDeep(context);
+            const pageCount = Math.ceil(
+                context.snapshotContext!.totalCount! / context.snapshotContext!.countPerPage!,
             );
-
-            context.returnedExtensions.snapResponse = { pagesIds };
+            const snapshotPages: SnapshotPage[] = [];
+            const pagesIds: string[] = Array(pageCount)
+                .fill(0)
+                .map(function generateUUIDAndCreateSnapshotPage() {
```
Comment (osher-sade) [inline src/plugins/snapshot/snapshot-listener.ts]: Declare this function outside of the current function, and tranform into an arrow function

File: src/server/polaris-server.ts (hunk 4 above)
Comment (osher-sade) [reply src/server/polaris-server.ts]: Decorator*

**Review by osher-sade — COMMENTED** [review]
Run npm update && npm audit fix

File: src/plugins/snapshot/snapshot-listener.ts (hunk 1 above)
Comment (DoctorVoid) [reply src/plugins/snapshot/snapshot-listener.ts]: to seperate main context which will return the snapresponse, from the context that will be used in page generation (which doesn't supposed to hold for example the snap response fields)

**Review by chenshoo — APPROVED** [review]

**Pull request merged**